Paintroid/src/main/java/org/catrobat/paintroid/command/implementation/DefaultCommandManager.kt

```diff
@@ -90,23 +90,37 @@ class DefaultCommandManager(
     }
 
     override fun undo() {
+        var success = true
         val command = undoCommandList.pop()
         redoCommandList.addFirst(command)
-
         var layerCount = layerModel.layerCount
         val currentCommandName = command.javaClass.simpleName
         val addLayerCommandRegex = AddLayerCommand::class.java.simpleName.toRegex()
         val mergeLayerCommandRegex = MergeLayersCommand::class.java.simpleName.toRegex()
 
+        var backupLayer: LayerContracts.Layer? = null
         if (currentCommandName.matches(addLayerCommandRegex)) {
             layerCount--
-            layerModel.removeLayerAt(0)
+            backupLayer = layerModel.getLayerAt(0)
+            success = layerModel.removeLayerAt(0)
         }
 
         val checkBoxes: MutableList<Boolean> = ArrayList(Collections.nCopies(layerCount, true))
 
         if (!currentCommandName.matches(mergeLayerCommandRegex)) {
-            backUpCheckBoxes(layerCount, checkBoxes)
+            success = if (!backUpCheckBoxes(layerCount, checkBoxes)) false else success
+        }
+
+        if (!success) {
+            backupLayer?.apply {
+                if (layerModel.getLayerIndexOf(this) != -1) {
```

Comment:
ah.. you're trying to check if the backupLayer is not present in the layerModel.layers, then you're adding it to the layerModel.layers, right 🤔? If yes, then I think the condition should be `layerModel.getLayerIndexOf(this) == -1`

Reply:
Good catch! The backup layer could get popped from the list above, if it was removed, we want to add it back to the original position at position 0!